pyomo/contrib/mindtpy/single_tree.py

```diff
@@ -657,256 +651,143 @@ def handle_lazy_subproblem_other_termination(
                 )
-                solve_data.primal_bound_improved = False
+                mindtpy_object.primal_bound_improved = False
                 return
-            if solve_data.dual_bound != solve_data.dual_bound_progress[0]:
-                main_mip, main_mip_results = solve_main(
-                    solve_data, config, regularization_problem=True
-                )
-                self.handle_lazy_regularization_problem(
-                    main_mip, main_mip_results, solve_data, config
-                )
+            if mindtpy_object.dual_bound != mindtpy_object.dual_bound_progress[0]:
+                mindtpy_object.add_regularization()
         if (
-            abs(solve_data.primal_bound - solve_data.dual_bound)
+            abs(mindtpy_object.primal_bound - mindtpy_object.dual_bound)
             <= config.absolute_bound_tolerance
         ):
             config.logger.info(
                 'MindtPy exiting on bound convergence. '
                 '|Primal Bound: {} - Dual Bound: {}| <= (absolute tolerance {})  \n'.format(
-                    solve_data.primal_bound,
-                    solve_data.dual_bound,
+                    mindtpy_object.primal_bound,
+                    mindtpy_object.dual_bound,
                     config.absolute_bound_tolerance,
                 )
             )
-            solve_data.results.solver.termination_condition = tc.optimal
+            mindtpy_object.results.solver.termination_condition = tc.optimal
             self.abort()
             return
 
         # check if the same integer combination is obtained.
-        solve_data.curr_int_sol = get_integer_solution(
-            solve_data.working_model, string_zero=True
+        mindtpy_object.curr_int_sol = get_integer_solution(
+            mindtpy_object.fixed_nlp, string_zero=True
         )
 
-        if solve_data.curr_int_sol in set(solve_data.integer_list):
+        if mindtpy_object.curr_int_sol in set(mindtpy_object.integer_list):
             config.logger.debug(
                 'This integer combination has been explored. '
                 'We will skip solving the Fixed-NLP subproblem.'
             )
-            solve_data.primal_bound_improved = False
+            mindtpy_object.primal_bound_improved = False
             if config.strategy == 'GOA':
                 if config.add_no_good_cuts:
                     var_values = list(
                         v.value
-                        for v in solve_data.working_model.MindtPy_utils.variable_list
+                        for v in mindtpy_object.working_model.MindtPy_utils.variable_list
                     )
-                    self.add_lazy_no_good_cuts(var_values, solve_data, config, opt)
-                return
-            elif config.strategy == 'OA':
+                    self.add_lazy_no_good_cuts(var_values, mindtpy_object, config, opt)
                 return
+            # TODO: check why this need to be commented. This previously worked well.
```

Comment:
What is happening here?

Reply:
In the single-tree implementation, the same integer combination could be obtained twice in the branch and bound process. Previously, if this happens, we will skip and avoid solving the same fixed-NLP subproblem again since the OA cuts have already been added as lazy constraints. This worked well. 
However, if lazy constraint callback is invoked during MIP start processing, the OA cuts are not actually added. We cannot skip and need to add them again in this case.
(I think the CPLEX group updated the documentation recently.)

Reference:
https://www.ibm.com/docs/en/icos/22.1.1?topic=SSSA5P_22.1.1/ilog.odms.cplex.help/refpythoncplex/html/cplex.callbacks.LazyConstraintCallback-class.htm
> Note:
> The lazy constraint callback may be invoked during MIP start processing. In that case get_solution_source returns mip_start_solution. When this value is returned some special considerations apply:
> 
> “
> MIP start processing occurs very early in the solution process. At this point no search tree is setup yet and there are no search tree nodes yet. Consequently, a lot of the callback methods that require a node context will fail in this situation.
> Lazy constraints separated when processing a MIP start will be discarded after that MIP start has been processed. This means that the callback may have to separate the same constraint again for the next MIP start or for a solution that is found later in the solution process.
> ”

Reply:
Understood, leave a comment here for future reference